### PR TITLE
resource/aws_route53_resolver_firewall_rule_group_association: Add new resource

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates:  #18520
Replaces: #18731

Output from acceptance testing:

#### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_'                          
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears (189.94s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection (203.66s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic (212.09s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags (240.66s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority (298.32s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name (388.28s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	391.720s
```

#### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_'  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears (224.56s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic (238.91s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection (259.86s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags (270.61s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority (327.57s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name (408.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	411.195s
```

### CODE REVIEW

File: aws/internal/service/route53resolver/waiter/waiter.go [hunk 1]
```diff
@@ -187,3 +196,58 @@ func FirewallDomainListDeleted(conn *route53resolver.Route53Resolver, firewallDo
 
 	return nil, err
 }
+
+// FirewallRuleGroupAssociationCreated waits for a FirewallRuleGroupAssociation to return COMPLETE
+func FirewallRuleGroupAssociationCreated(conn *route53resolver.Route53Resolver, firewallRuleGroupAssociationId string) (*route53resolver.FirewallRuleGroupAssociation, error) {
+	stateConf := &resource.StateChangeConf{
+		// TODO: What should this be?
```
Comment (bflad) [inline aws/internal/service/route53resolver/waiter/waiter.go:203]: Is this still a TODO?

```suggestion
```

File: aws/resource_aws_route53_resolver_firewall_rule_group_association.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"vpc_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateRoute53ResolverName,
+			},
```
Comment (bflad) [inline aws/resource_aws_route53_resolver_firewall_rule_group_association.go:63]: Nit: Should this use a different validation?

File: aws/resource_aws_route53_resolver_firewall_rule_group_association.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+func resourceAwsRoute53ResolverFirewallRuleGroupAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).route53resolverconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	ruleGroupAssociation, err := finder.FirewallRuleGroupAssociationByID(conn, d.Id())
+
+	if isAWSErr(err, route53resolver.ErrCodeResourceNotFoundException, "") {
```
Comment (bflad) [inline aws/resource_aws_route53_resolver_firewall_rule_group_association.go:115]: Reference: https://github.com/hashicorp/terraform-provider-aws/issues/16796

```suggestion
	if !d.IsNewResource() && isAWSErr(err, route53resolver.ErrCodeResourceNotFoundException, "") {
```

File: aws/resource_aws_route53_resolver_firewall_rule_group_association.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+	if err != nil {
+		return fmt.Errorf("error getting Route 53 Resolver DNS Firewall rule group association (%s): %w", d.Id(), err)
+	}
+
+	if ruleGroupAssociation == nil {
+		log.Printf("[WARN] Route 53 Resolver DNS Firewall rule group association (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
```
Comment (bflad) [inline aws/resource_aws_route53_resolver_firewall_rule_group_association.go:129]: Reference: https://github.com/hashicorp/terraform-provider-aws/issues/16796

```suggestion
	if ruleGroupAssociation == nil {
		if d.IsNewResource() {
			return fmt.Errorf("error getting Route 53 Resolver DNS Firewall rule group association (%s): not found after creation", d.Id())
		}

		log.Printf("[WARN] Route 53 Resolver DNS Firewall rule group association (%s) not found, removing from state", d.Id())
		d.SetId("")
		return nil
	}
```

**Review by bflad — APPROVED** [review]
A few minor things, otherwise looks good to me. 🚀

File: aws/resource_aws_route53_resolver_firewall_rule_group_association.go (hunk 2 above)
Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_firewall_rule_group_association.go:63]: Good spot. I'll remove the validation on this field.

Comment (ewbankkit) [reply aws/resource_aws_route53_resolver_firewall_rule_group_association.go]: Done (plus removed the validation on `firewall_rule_group_id`).

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_'  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_ -timeout 180m
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== RUN   TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== PAUSE TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection
=== CONT  TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_basic (186.63s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_disappears (192.39s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_mutationProtection (207.96s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_tags (228.00s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_priority (294.59s)
--- PASS: TestAccAWSRoute53ResolverFirewallRuleGroupAssociation_name (354.24s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	357.279s
```

**Pull request merged**